presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -314,18 +315,50 @@ public FileStatus getFileStatus(Path path)
         }
 
         return new FileStatus(
-                getObjectSize(metadata),
+                getObjectSize(metadata, path),
                 false,
                 1,
                 BLOCK_SIZE.toBytes(),
                 lastModifiedTime(metadata),
                 qualifiedPath(path));
     }
 
-    private static long getObjectSize(ObjectMetadata metadata)
+    private long getObjectSize(ObjectMetadata metadata, Path path) throws IOException
     {
         String length = metadata.getUserMetadata().get(UNENCRYPTED_CONTENT_LENGTH);
-        return (length != null) ? Long.parseLong(length) : metadata.getContentLength();
+
+        if (length != null) {
+            return Long.parseLong(length);
+        }
+
+        // check if file has KMS encryption
+        String cryptoAlg = metadata.getUserMetadata().get(CRYPTO_KEYWRAP_ALGORITHM);
+        if (cryptoAlg != null && cryptoAlg.equalsIgnoreCase("kms")) {
+            // KMS client-side encryption
+            long decryptedLength = metadata.getContentLength();
+
+            // load file and determine actual length
+            FSDataInputStream inputStream = this.open(path);
+
+            // read last 25 bytes of file to help determine length
+            int eofOffset = 25;
```

Comment:
My goal is to only read a small amount of the S3 file to minimize unnecessary calls to S3 and KMS.  During my investigation the differences between the encrypted KMS file and the final decrypted version were in the 4-15 bytes range.  I have yet to hit an issue with this value during my testing of actual files but am open to bumping this up to ensure we capture all cases.  Also open to alternative approaches if anyone has thoughts.